keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,955 @@
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes to extend the Job API by adding an optional `SetConditionReason` field to `PodFailurePolicyRule`, which if specified, would be included as the reason in the `JobFailed` condition upon Job failure triggered by a `PodFailurePolicy`.
+
+## Motivation
+
+Higher level APIs, such as [JobSet](https://sigs.k8s.io/jobset), use the Job API as a building block for orchestrating large, distributed workloads.
+These higher level APIs need to be able to distinguish between different types of Job failures in order to make informed decisions about how to react
+to them. Currently, no mechanism exists in the Job API to propagate granular failure reason information (e.g., container exit codes) up to be 
+programmatically consumed by higher level software managing Jobs. A `PodFailurePolicy` can be configured to add a condition reason of `PodFailurePolicy`
+to the `JobFailed` condition added to the Job when it fails, but different pod failure policies targeting different container exit codes all use the
+same condition reason of `PodFailurePolicy`. This prevents higher level APIs like JobSet from distinguishing them and being able to take different
+actions depending on the type of Job failure that occurred.
+
+For a concrete use case, see the JobSet [Configurable Failury Policy KEP](https://github.com/kubernetes-sigs/jobset/pull/381) which illuminated the need for more granular pod failure policy reasons.
+
+### Goals
+
+For pod failure policies to be able communicate different failure types to higher level APIs.
+
+### Non-Goals
+
+- Modifying PodFailurePolicy behavior
+- The Job controller using this new field for any purpose not explicitly defined in the proposal.
+
+## Proposal
+
+The proposal is to add an optional `SetConditionReason` field to the `PodFailurePolicyRule` that is used
+to set the `Reason` field on the `JobFailed` condition when a Job fails because of a pod failure policy rule match.
+
+If unset, it will default to `PodFailurePolicy`, which is the current [reason](https://sourcegraph.com/github.com/kubernetes/kubernetes@6a4e93e776a35d14a61244185c848c3b5832621c/-/blob/staging/src/k8s.io/api/batch/v1/types.go?L542) the Job
+controller uses when a PodFailurePolicy triggers a Job failure.
+
+When a `PodFailurePolicyRule` matches a pod failure and the `Action` is `FailJob`, the Job
+controller will add the reason defined in the `SetConditionReason` field to the JobFailed [condition](https://sourcegraph.com/github.com/kubernetes/kubernetes@6a4e93e776a35d14a61244185c848c3b5832621c/-/blob/pkg/controller/job/job_controller.go?L816) added
+to the Job.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
```

Comment:
nit: JobSet is still an alpha project so it may be worth thinking of more general cases than just for one project.

There could be use for a Job to add a more detailed condition so that higher level controllers can act on that (aka jobSet ;)) but I just think we don't want to make it sound like the only reason we are adding this is because we would find it useful in JobSet.

Reply:
The motivation is not JobSet specifically, the motivation is allowing higher-order apis to understand failure reasons and react based on that. As a k8s project, we are trying to encourage composition of features (in this case building advanced capabilities from upstream building blocks).

The story here gives a concrete example of that, I actually think it is very appropriate they way it is described.

Are you asking for another story to discuss another hypothetical controller?

Reply:
Maybe a new story focusing on why this is useful outside of JobSet. Composition of features is a good point and I hope the feature can go in. I can see the usefulness in building objects on top of the Job and this could be necessary for UX for detecting failures due to PodFailurePolicy.

Reply:
So my point was mostly this is useful to other users of the Job. The stories are framed in a way to say that this is useful to a JobSet but it isn't clear if it would be useful to other projects with this stories. JobSet is an example but its not the reason why we are implementing this. We just discovered that this would be useful to build more robust failure/success policies using PodFailurePolicy and control this in a smarter way.

Reply:
I would just update the motivation to stress the composition point. I prefer the stories to be concrete and less hypothetical.

Reply:
It seems there is some conflicting opinions on this; in the original KEP I had a more generic example before not involving JobSet, but got feedback that I should remove it. 

As a middle ground, I have updated the motivation section to start by talking more generally about higher level APIs built via a composition of features, using K8s primitive APIs as building blocks for implementing more advanced functionality, then describing why higher level APIs using the Job API as a building block need to be able to distinguish between different types of Job failures with more granularity in order to react to them accordingly.

Reply:
That looks good to me @danielvegamyhre!